lib/Runtime/Library/JavascriptGeneratorFunction.cpp

```diff
@@ -526,17 +525,62 @@ namespace Js
     }
 
 #if ENABLE_TTD
+
+    void JavascriptGeneratorFunction::MarkVisitKindSpecificPtrs(TTD::SnapshotExtractor* extractor)
+    {
+        if (this->scriptFunction != nullptr)
+        {
+            extractor->MarkVisitVar(this->scriptFunction);
+        }
+    }
+
     TTD::NSSnapObjects::SnapObjectType JavascriptGeneratorFunction::GetSnapTag_TTD() const
     {
         return TTD::NSSnapObjects::SnapObjectType::SnapGeneratorFunction;
     }
 
     void JavascriptGeneratorFunction::ExtractSnapObjectDataInto(TTD::NSSnapObjects::SnapObject* objData, TTD::SlabAllocator& alloc)
     {
-        TTD::NSSnapObjects::SnapGeneratorFunctionInfo* fi = alloc.SlabAllocateStruct<TTD::NSSnapObjects::SnapGeneratorFunctionInfo>();
-        fi->scriptFunction = TTD_CONVERT_VAR_TO_PTR_ID(this->scriptFunction);
-        fi->isAnonymousFunction = this->scriptFunction->IsAnonymousFunction();
-        TTD::NSSnapObjects::StdExtractSetKindSpecificInfo<TTD::NSSnapObjects::SnapGeneratorFunctionInfo*, TTD::NSSnapObjects::SnapObjectType::SnapGeneratorFunction>(objData, fi);
+        TTD::NSSnapObjects::SnapGeneratorFunctionInfo* fi = nullptr;
+        uint32 depCount = 0;
+        TTD_PTR_ID* depArray = nullptr;
+
+        this->CreateSnapObjectInfo(alloc, &fi, &depArray, &depCount);
+
+        if (depCount == 0)
+        {
+            TTD::NSSnapObjects::StdExtractSetKindSpecificInfo<TTD::NSSnapObjects::SnapGeneratorFunctionInfo*, TTD::NSSnapObjects::SnapObjectType::SnapGeneratorFunction>(objData, fi);
+        }
+        else
+        {
+            TTDAssert(depArray != nullptr, "depArray should be non-null if depCount is > 0");
+            TTD::NSSnapObjects::StdExtractSetKindSpecificInfo<TTD::NSSnapObjects::SnapGeneratorFunctionInfo*, TTD::NSSnapObjects::SnapObjectType::SnapGeneratorFunction>(objData, fi, alloc, depCount, depArray);
+        }
+    }
+
+    void JavascriptGeneratorFunction::CreateSnapObjectInfo(TTD::SlabAllocator& alloc, _Out_ TTD::NSSnapObjects::SnapGeneratorFunctionInfo** info, _Out_ TTD_PTR_ID** depArray, _Out_ uint32* depCount)
+    {
+        *info = alloc.SlabAllocateStruct<TTD::NSSnapObjects::SnapGeneratorFunctionInfo>();
+        (*info)->scriptFunction = TTD_CONVERT_VAR_TO_PTR_ID(this->scriptFunction);
+        (*info)->isAnonymousFunction = this->scriptFunction->IsAnonymousFunction();
+
+        const uint32 maxDeps = 1;
+        *depCount = 0;
+        *depArray = alloc.SlabReserveArraySpace<TTD_PTR_ID>(maxDeps);
```

Comment:
I'm a little confused by this code here; it looks like the code is set up to handle some variable number of entries, but only ever have at most 1, for `this->scriptFunction`. What is the reasoning behind having an array and a count of dependencies?

Reply:
Good feedback.  I simplified this a bit.  The reason we use an array & count here is that is what we need to pass into existing TTD APIs.